src/Modal.tsx

```diff
@@ -0,0 +1,527 @@
+      manager: providedManager,
+      container: containerRef,
+      containerClassName,
+      onShow,
+      onHide = () => {},
+
+      onExit,
+      onExited,
+      onExiting,
+      onEnter,
+      onEntering,
+      onEntered,
+
+      ...rest
+    }: ModalProps,
+    ref,
+  ) => {
+    // const [manager] = useModalManager(providedManager);
+    const container = useWaitForDOMRef(containerRef);
+    const modal = useModalManager(providedManager);
+
+    const isMounted = useMounted();
+    const prevShow = usePrevious(show);
+    const [exited, setExited] = useState(!show);
+    const lastFocusRef = useRef<HTMLElement | null>(null);
+
+    useImperativeHandle(ref, () => modal, [modal]);
+
+    if (canUseDOM && !prevShow && show) {
+      lastFocusRef.current = activeElement() as HTMLElement;
+    }
+
+    if (!transition && !show && !exited) {
+      setExited(true);
+    } else if (show && exited) {
+      setExited(false);
+    }
+
+    const handleShow = useEventCallback(() => {
+      modal.add(container!, containerClassName);
+
+      removeKeydownListenerRef.current = listen(
+        document as any,
+        'keydown',
+        handleDocumentKeyDown,
+      );
+
+      removeFocusListenerRef.current = listen(
+        document as any,
+        'focus',
+        // the timeout is necessary b/c this will run before the new modal is mounted
+        // and so steals focus from it
+        () => setTimeout(handleEnforceFocus),
+        true,
+      );
+
+      if (onShow) {
+        onShow();
+      }
+
+      // autofocus after onShow to not trigger a focus event for previous
+      // modals before this one is shown.
+      if (autoFocus) {
+        const currentActiveElement = activeElement(document) as HTMLElement;
+
+        if (
+          modal.dialog &&
+          currentActiveElement &&
+          !contains(modal.dialog, currentActiveElement)
+        ) {
+          lastFocusRef.current = currentActiveElement;
+          modal.dialog.focus();
+        }
+      }
+    });
+
+    const handleHide = useEventCallback(() => {
+      modal.remove();
+
+      removeKeydownListenerRef.current?.();
+      removeFocusListenerRef.current?.();
+
+      if (restoreFocus) {
+        // Support: <=IE11 doesn't support `focus()` on svg elements (RB: #917)
+        lastFocusRef.current?.focus?.(restoreFocusOptions);
+        lastFocusRef.current = null;
+      }
+    });
+
+    // Show logic when:
+    //  - show is `true` _and_ `container` has resolved
+    useEffect(() => {
+      if (!show || !container) return;
+
+      handleShow();
+    }, [show, container, /* show never change: */ handleShow]);
+
+    // Hide cleanup logic when:
+    //  - `exited` switches to true
+    //  - component unmounts;
+    useEffect(() => {
+      if (!exited) return;
+
+      handleHide();
+    }, [exited, handleHide]);
+
+    useWillUnmount(() => {
+      handleHide();
+    });
```

Comment:
is there no good way to write this as a single effect? e.g. calculate a shown value, and then have an effect block?

```js
if (show && container && !foo) {
  setFoo(true);
} else if (exited) {
  setFoo(false);
}

useEffect(() => {
  if (!foo) {
    return undefined;
  }

  handleShow();
  return handleHide;
}, [foo, handleShow, handleHide]);
```

Reply:
ah maybe, wasn't sure if it helped us much tho. I guess it would simplify it nicely. I can try

Reply:
imma try that as a follow up